# Ejercicios Basicos dataframe



In [134]:
import pandas as pd
import numpy as np

### 1. Carga el dataframe con nombre “EjerciciosPracticos_datos_Dataframes.csv”
El dataframe viene en formato csv con columnas "rate", "country", "user_id", "source" y "topic". Nota: Recomiendo primero abrir el csv (con excel o con el bloc de notas) y observar el formato del archivo. Tendréis que modificar argumentos de la función read_csv para leerlo con éxito.

In [164]:
datos = pd.read_csv("EjerciciosPracticos_datos_Dataframes.csv", names = ["rate", "country", "user_id", "source", "topic"])

### 2. Imprime

    a.Imprime el dataframe entero por pantalla
    
    b.Imprime las 5 primeras filas
    
    c.Imprime las 5 últimas filas
    
    d.Imprime de la fila 10 a la 15
    
    e.Imprime el número de filas y de columnas
    
    f.Imprime la columna source
    
    g.Imprime las columnas source y topic a la vez
    
    h.Imprime los 5 primeros registros de las columnas source y topic a la vez
    
    i.Imprime los registros con topic = Asia
    
    j.Imprime los registros con topic = Asia y source = SEO
    
    k.Imprime las columnas rate y country de los registros con topic = Asia y source = SEO
    
    l.Imprime solo los 5 primeros registros del caso anterior
    
    m.Imprime el número de filas y columnas del caso en ejercicio en k


In [165]:
# a)
datos


rate    country     user_id   source          topic
0     0.00  country_7  2458151261      SEO  North America
1     0.00  country_7  2458151262      SEO  South America
2     0.00  country_7  2458151263  AdWords         Africa
3     0.00  country_7  2458151264  AdWords         Europe
4     0.00  country_8  2458151265   Reddit  North America
...    ...        ...         ...      ...            ...
1790  0.98  country_2  2458153051  AdWords  North America
1791  0.98  country_8  2458153052      SEO           Asia
1792  0.98  country_6  2458153053   Reddit           Asia
1793  0.98  country_7  2458153054  AdWords         Europe
1794  0.98  country_5  2458153055   Reddit           Asia

[1795 rows x 5 columns]

In [137]:
# b)
datos.head(5)

rate    country     user_id   source          topic
0   0.0  country_7  2458151261      SEO  North America
1   0.0  country_7  2458151262      SEO  South America
2   0.0  country_7  2458151263  AdWords         Africa
3   0.0  country_7  2458151264  AdWords         Europe
4   0.0  country_8  2458151265   Reddit  North America

In [138]:
# c)
datos.tail(5)

rate    country     user_id   source          topic
1790  0.98  country_2  2458153051  AdWords  North America
1791  0.98  country_8  2458153052      SEO           Asia
1792  0.98  country_6  2458153053   Reddit           Asia
1793  0.98  country_7  2458153054  AdWords         Europe
1794  0.98  country_5  2458153055   Reddit           Asia

In [139]:
# d)
datos[10:16]

rate    country     user_id   source          topic
10  0.01  country_5  2458151271  AdWords           Asia
11  0.01  country_7  2458151272      SEO      Australia
12  0.01  country_5  2458151273   Reddit           Asia
13  0.01  country_2  2458151274   Reddit         Europe
14  0.01  country_7  2458151275   Reddit         Africa
15  0.01  country_7  2458151276      SEO  North America

In [140]:
# e)
datos.shape

(1795, 5)

In [141]:
# f)
datos.source

0           SEO
1           SEO
2       AdWords
3       AdWords
4        Reddit
         ...   
1790    AdWords
1791        SEO
1792     Reddit
1793    AdWords
1794     Reddit
Name: source, Length: 1795, dtype: object

In [166]:
# g)
datos[["source", "topic"]]

source          topic
0         SEO  North America
1         SEO  South America
2     AdWords         Africa
3     AdWords         Europe
4      Reddit  North America
...       ...            ...
1790  AdWords  North America
1791      SEO           Asia
1792   Reddit           Asia
1793  AdWords         Europe
1794   Reddit           Asia

[1795 rows x 2 columns]

In [167]:
# h)
datos[["source", "topic"]].head(5)

source          topic
0      SEO  North America
1      SEO  South America
2  AdWords         Africa
3  AdWords         Europe
4   Reddit  North America

In [144]:
# i)
datos[datos.topic=="Asia"]

rate    country     user_id   source topic
10    0.01  country_5  2458151271  AdWords  Asia
12    0.01  country_5  2458151273   Reddit  Asia
17    0.02  country_2  2458151278   Reddit  Asia
19    0.02  country_2  2458151280   Reddit  Asia
20    0.02  country_2  2458151281   Reddit  Asia
...    ...        ...         ...      ...   ...
1787  0.98  country_2  2458153048   Reddit  Asia
1789  0.98  country_4  2458153050  AdWords  Asia
1791  0.98  country_8  2458153052      SEO  Asia
1792  0.98  country_6  2458153053   Reddit  Asia
1794  0.98  country_5  2458153055   Reddit  Asia

[667 rows x 5 columns]

In [168]:
# j)
datos[(datos.topic=="Asia") & (datos.source=="SEO")]

rate    country     user_id source topic
41    0.04  country_5  2458151302    SEO  Asia
62    0.06  country_6  2458151323    SEO  Asia
98    0.10  country_4  2458151359    SEO  Asia
207   0.21  country_2  2458151468    SEO  Asia
228   0.23  country_6  2458151489    SEO  Asia
295   0.29  country_8  2458151556    SEO  Asia
324   0.32  country_5  2458151585    SEO  Asia
358   0.35  country_2  2458151619    SEO  Asia
421   0.41  country_2  2458151682    SEO  Asia
491   0.47  country_2  2458151752    SEO  Asia
648   0.60  country_5  2458151909    SEO  Asia
650   0.61  country_2  2458151911    SEO  Asia
692   0.64  country_2  2458151953    SEO  Asia
709   0.65  country_2  2458151970    SEO  Asia
746   0.68  country_8  2458152007    SEO  Asia
827   0.74  country_5  2458152088    SEO  Asia
897   0.78  country_7  2458152158    SEO  Asia
1054  0.87  country_5  2458152315    SEO  Asia
1140  0.91  country_6  2458152401    SEO  Asia
1154  0.91  country_5  2458152415    SEO  Asia
1168  0.92  country_5  2458152429    SEO  Asia
1169  0.92  country_8  2458152430    SEO  Asia
1232  0.94  country_2  2458152493    SEO  Asia
1257  0.95  country_1  2458152518    SEO  Asia
1317  0.97  country_8  2458152578    SEO  Asia
1333  0.97  country_6  2458152594    SEO  Asia
1357  0.98  country_8  2458152618    SEO  Asia
1527  1.00  country_7  2458152788    SEO  Asia
1528  1.00  country_4  2458152789    SEO  Asia
1601  1.00  country_2  2458152862    SEO  Asia
1675  0.99  country_5  2458152936    SEO  Asia
1736  0.99  country_5  2458152997    SEO  Asia
1740  0.99  country_7  2458153001    SEO  Asia
1791  0.98  country_8  2458153052    SEO  Asia

In [169]:
# k)
k = ["rate", "country"]
datos[(datos.topic=="Asia") & (datos.source=="SEO")][k]

rate    country
41    0.04  country_5
62    0.06  country_6
98    0.10  country_4
207   0.21  country_2
228   0.23  country_6
295   0.29  country_8
324   0.32  country_5
358   0.35  country_2
421   0.41  country_2
491   0.47  country_2
648   0.60  country_5
650   0.61  country_2
692   0.64  country_2
709   0.65  country_2
746   0.68  country_8
827   0.74  country_5
897   0.78  country_7
1054  0.87  country_5
1140  0.91  country_6
1154  0.91  country_5
1168  0.92  country_5
1169  0.92  country_8
1232  0.94  country_2
1257  0.95  country_1
1317  0.97  country_8
1333  0.97  country_6
1357  0.98  country_8
1527  1.00  country_7
1528  1.00  country_4
1601  1.00  country_2
1675  0.99  country_5
1736  0.99  country_5
1740  0.99  country_7
1791  0.98  country_8

In [174]:
# l)

k = ["rate", "country"]
datos[(datos.topic=="Asia") & (datos.source=="SEO")][k].head(5)

rate    country
41   0.04  country_5
62   0.06  country_6
98   0.10  country_4
207  0.21  country_2
228  0.23  country_6

In [175]:
# m)
k = ["rate", "country"]
datos[(datos.topic=="Asia") & (datos.source=="SEO")][k].shape

(34, 2)

### 3. Cálculos básicos:

a. ¿Cuál es el máximo y el mínimo de la variable user_id?

b. ¿Cuál es la media y la sd de la variable rate?

c. ¿Cuál es el percentil 25 y 75 de la variable rate?

d. ¿Cuál es el valor más frecuente de la variable topic? [R: Asia con 667 casos]

e ¿Cuál es el country que obtiene un rate medio mayor? [R: Country_4 con 0.72 de media]

In [176]:
datos.describe()

rate       user_id
count  1795.000000  1.795000e+03
mean      0.680763  2.458152e+09
std       0.311360  5.183162e+02
min       0.000000  2.458151e+09
25%       0.430000  2.458152e+09
50%       0.780000  2.458152e+09
75%       0.970000  2.458153e+09
max       1.000000  2.458153e+09

In [177]:
# a)
print(datos.user_id.min()," , ",datos.user_id.max())

2458151261  ,  2458153055


In [178]:
# b)
print(datos.rate.mean()," , ",datos.rate.std())

0.6807632311977716  ,  0.3113600316793178


In [179]:
# c)
print(datos.rate.quantile(0.25)," , ",datos.rate.quantile(0.75))

0.43  ,  0.97


In [180]:
# d)
datos.topic.value_counts()

topic
Asia             667
North America    326
Europe           322
South America    234
Africa           125
Australia        121
Name: count, dtype: int64

In [181]:
# e)¿Cuál es el country que obtiene un rate medio mayor? [R: Country_4 con 0.72 de media]
mayor= ["error", -1]
for i in datos.country.unique():
    j = datos[datos.country==i].rate.mean()
    if mayor[1] < j:
        mayor = [i,j]
mayor

['country_4', 0.7233333333333333]


### 4. Subsets:
a. ¿Cuál es el percentil 50 de la variable rate en los registros con topic = Asia? [R: 0.77]

b. ¿Cuál es el valor de la variable country más frecuente en los registros con topic = Asia? [R: country_2 con 179 registros]

c. ¿Cuál es el valor de la variable country más frecuente en los registros con topic = Asia y source = SEO? [R: country_2 y country_5 con 9 casos cada uno]

d. ¿Cuál es el user_id de los tres registros con menor rate de topic = Asia y source = SEO? [R: 2458151302, 2458151323 y 2458151359]

In [182]:
# a)
datos.rate[datos.topic=="Asia"].quantile(0.5)

0.77

In [183]:
# b)
datos.country[datos.topic=="Asia"].value_counts().head(1)

country
country_2    179
Name: count, dtype: int64

In [184]:
# c)
datos.country[(datos.topic=="Asia") & (datos.source=="SEO")].value_counts().head(2)

country
country_5    9
country_2    9
Name: count, dtype: int64

In [185]:
# d)
form=(datos.topic=="Asia") & (datos.source=="SEO")
out=["user_id"]
datos[form].sort_values(by="rate")[out].head(3)

user_id
41  2458151302
62  2458151323
98  2458151359

### 5. Operaciones por columnas:
a. Crea la variable rate2 como la raíz cuadrada de la variable rate.

b. Crea la variable randUni a partir de generar valores aleatorios de una distribución uniforme[0,1]

c. Crea la variable topic_country concatenando las columnas topic y country

d. Ejecuta y entiende las siguientes líneas de código:
datos["country_num"] = datos.country.str.slice(8, 9).astype(int)
datos["country"] = "pais_" + datos["country_num"].astype(str)

e. Elimina definitivamente del dataframe las columnas rate2, topic_country y country_num.

f. Sustituye los valores de AdWords, Reddit y SEO de la columna source por sus iniciales A, R y S.

g. Crea la variable muestra y que esta tome valor 1 para los valores de randUni menores a 0.7 y 0 para el resto.

h. Crea la variable submuestra y que esta tome valor 1 para los valores que muestra = 1 y que source = S.

In [201]:
# a)
datos["rate2"] = datos.rate**(1/2)
datos

rate country     user_id   source          topic   randUni     rate2
0     0.00  pais_7  2458151261      SEO  North America  0.987215  0.000000
1     0.00  pais_7  2458151262      SEO  South America  0.926764  0.000000
2     0.00  pais_7  2458151263  AdWords         Africa  0.373135  0.000000
3     0.00  pais_7  2458151264  AdWords         Europe  0.896796  0.000000
4     0.00  pais_8  2458151265   Reddit  North America  0.292921  0.000000
...    ...     ...         ...      ...            ...       ...       ...
1790  0.98  pais_2  2458153051  AdWords  North America  0.029511  0.989949
1791  0.98  pais_8  2458153052      SEO           Asia  0.165412  0.989949
1792  0.98  pais_6  2458153053   Reddit           Asia  0.276657  0.989949
1793  0.98  pais_7  2458153054  AdWords         Europe  0.167289  0.989949
1794  0.98  pais_5  2458153055   Reddit           Asia  0.125263  0.989949

[1795 rows x 7 columns]

In [202]:
# b)
datos["randUni"] = np.random.uniform(size=datos.shape[0])
datos.head()

rate country     user_id   source          topic   randUni  rate2
0   0.0  pais_7  2458151261      SEO  North America  0.703307    0.0
1   0.0  pais_7  2458151262      SEO  South America  0.339866    0.0
2   0.0  pais_7  2458151263  AdWords         Africa  0.047578    0.0
3   0.0  pais_7  2458151264  AdWords         Europe  0.151442    0.0
4   0.0  pais_8  2458151265   Reddit  North America  0.199570    0.0

In [203]:
# c)
datos["topic_country"] = datos.topic + "_" + datos.country
datos

rate country     user_id   source          topic   randUni     rate2  \
0     0.00  pais_7  2458151261      SEO  North America  0.703307  0.000000   
1     0.00  pais_7  2458151262      SEO  South America  0.339866  0.000000   
2     0.00  pais_7  2458151263  AdWords         Africa  0.047578  0.000000   
3     0.00  pais_7  2458151264  AdWords         Europe  0.151442  0.000000   
4     0.00  pais_8  2458151265   Reddit  North America  0.199570  0.000000   
...    ...     ...         ...      ...            ...       ...       ...   
1790  0.98  pais_2  2458153051  AdWords  North America  0.007689  0.989949   
1791  0.98  pais_8  2458153052      SEO           Asia  0.896861  0.989949   
1792  0.98  pais_6  2458153053   Reddit           Asia  0.566042  0.989949   
1793  0.98  pais_7  2458153054  AdWords         Europe  0.711321  0.989949   
1794  0.98  pais_5  2458153055   Reddit           Asia  0.066084  0.989949   

             topic_country  
0     North America_pais_7  
1     South America_pais_7  
2            Africa_pais_7  
3            Europe_pais_7  
4     North America_pais_8  
...                    ...  
1790  North America_pais_2  
1791           Asia_pais_8  
1792           Asia_pais_6  
1793         Europe_pais_7  
1794           Asia_pais_5  

[1795 rows x 8 columns]

d) Ejecuta y entiende las siguientes líneas de código:

    datos["country_num"] = datos.country.str.slice(8,9).astype(int)
    crea "country_num", y le asigana el numero de country, como tipo int

    datos["country"] = "pais_" + datos["country_num"].astype(str)
    reescribe la columna countri pasandle como par="pais_", y el numeor de antes


In [206]:
datos["country_num"] = datos.country.str.slice(8,9).astype(int)
datos["country"] = "pais_" + datos["country_num"].astype(str)

ValueError: invalid literal for int() with base 10: ''

In [207]:
# e) rate2, topic_country y country_num.
del datos["rate2"]
del datos["topic_country"]
del datos["country_num"]
datos


KeyError: 'rate2'

In [213]:
# f) Sustituye los valores de AdWords, Reddit y SEO de la columna source 
#    por sus iniciales A, R y S.
datos.source = datos.source.replace(["AdWords", "Reddit", "SEO"],["A","R","S"])

In [216]:
# g)Crea la variable muestra y que esta tome valor 1 para los valores de randUni menores a 0.7 y 0 para el resto.
datos["muestra"] = 0
datos.muestra[datos.randUni < 0.7] = 1
datos

# csv.loc[csv["Largo Sepalo"]>=5, "Largo Sepalo"] = 5.5

C:\Users\garch\AppData\Local\Temp\ipykernel_11868\996979073.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  datos.muestra[datos.randUni < 0.7] = 1


rate country     user_id source          topic   randUni  muestra
0     0.00  pais_7  2458151261      S  North America  0.703307        0
1     0.00  pais_7  2458151262      S  South America  0.339866        1
2     0.00  pais_7  2458151263      A         Africa  0.047578        1
3     0.00  pais_7  2458151264      A         Europe  0.151442        1
4     0.00  pais_8  2458151265      R  North America  0.199570        1
...    ...     ...         ...    ...            ...       ...      ...
1790  0.98  pais_2  2458153051      A  North America  0.007689        1
1791  0.98  pais_8  2458153052      S           Asia  0.896861        0
1792  0.98  pais_6  2458153053      R           Asia  0.566042        1
1793  0.98  pais_7  2458153054      A         Europe  0.711321        0
1794  0.98  pais_5  2458153055      R           Asia  0.066084        1

[1795 rows x 7 columns]

In [218]:
# h)Crea la variable submuestra y que esta tome valor 1 para los valores que muestra = 1 y que source = S.
datos["submuestra"] = 0
datos.submuestra[(datos.muestra == 1) & (datos.source == "S")] = 1
datos

C:\Users\garch\AppData\Local\Temp\ipykernel_11868\865224376.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  datos.submuestra[(datos.muestra == 1) & (datos.source == "S")] = 1


rate country     user_id source          topic   randUni  muestra  \
0     0.00  pais_7  2458151261      S  North America  0.703307        0   
1     0.00  pais_7  2458151262      S  South America  0.339866        1   
2     0.00  pais_7  2458151263      A         Africa  0.047578        1   
3     0.00  pais_7  2458151264      A         Europe  0.151442        1   
4     0.00  pais_8  2458151265      R  North America  0.199570        1   
...    ...     ...         ...    ...            ...       ...      ...   
1790  0.98  pais_2  2458153051      A  North America  0.007689        1   
1791  0.98  pais_8  2458153052      S           Asia  0.896861        0   
1792  0.98  pais_6  2458153053      R           Asia  0.566042        1   
1793  0.98  pais_7  2458153054      A         Europe  0.711321        0   
1794  0.98  pais_5  2458153055      R           Asia  0.066084        1   

      submuestra  
0              0  
1              1  
2              0  
3              0  
4              0  
...          ...  
1790           0  
1791           0  
1792           0  
1793           0  
1794           0  

[1795 rows x 8 columns]

### 6. Define un nuevo dataframe con las siguientes columnas: topic y hemisphere


https://cdn.discordapp.com/attachments/1214178634961784913/1244664694850060379/image.png?ex=6655efd4&is=66549e54&hm=87b5f222a7988ad872a37b3bcc13813028881561f312fbf7bcdcc96ffb461d97&

In [226]:
df = pd.DataFrame({"topic":["Africa","Australia","Asia","Europe","North America", "South America"],
                    "hemisphere":["South","South","North", "North", "North","South"]})
df

topic hemisphere
0         Africa      South
1      Australia      South
2           Asia      North
3         Europe      North
4  North America      North
5  South America      South

### 7. Incorpora el campo hemisphere al dataframe inicial a traves del campo común topic.
¿Cuántos registros hay con hemisphere north y cuantos south?


In [227]:
pd.merge(left=datos, right=df, on = "topic")

rate country     user_id source          topic   randUni  muestra  \
0     0.00  pais_7  2458151261      S  North America  0.703307        0   
1     0.00  pais_8  2458151265      R  North America  0.199570        1   
2     0.00  pais_6  2458151266      R  North America  0.856609        0   
3     0.01  pais_7  2458151269      A  North America  0.086956        1   
4     0.01  pais_5  2458151270      R  North America  0.643028        1   
...    ...     ...         ...    ...            ...       ...      ...   
1790  0.99  pais_2  2458152965      R      Australia  0.132804        1   
1791  0.99  pais_5  2458152966      R      Australia  0.070060        1   
1792  0.99  pais_7  2458152978      A      Australia  0.207989        1   
1793  0.99  pais_2  2458152981      R      Australia  0.324975        1   
1794  0.98  pais_5  2458153030      A      Australia  0.125185        1   

      submuestra hemisphere  
0              0      North  
1              0      North  
2              0      North  
3              0      North  
4              0      North  
...          ...        ...  
1790           0      South  
1791           0      South  
1792           0      South  
1793           0      South  
1794           0      South  

[1795 rows x 9 columns]